src/libcore/mem.rs

```diff
@@ -448,17 +448,29 @@ pub unsafe fn uninitialized<T>() -> T {
 pub fn swap<T>(x: &mut T, y: &mut T) {
     unsafe {
         // Give ourselves some scratch space to work with
-        let mut t: T = uninitialized();
+        let mut t: [u8; 16] = uninitialized();
 
-        // Perform the swap, `&mut` pointers never alias
-        ptr::copy_nonoverlapping(&*x, &mut t, 1);
-        ptr::copy_nonoverlapping(&*y, x, 1);
-        ptr::copy_nonoverlapping(&t, y, 1);
+        let x = x as *mut T as *mut u8;
+        let y = y as *mut T as *mut u8;
+        let t = &mut t as *mut _ as *mut u8;
 
-        // y and t now point to the same thing, but we need to completely
-        // forget `t` because we do not want to run the destructor for `T`
-        // on its value, which is still owned somewhere outside this function.
-        forget(t);
+        // can't use a for loop as the `range` impl calls `mem::swap` recursively
+        let len = size_of::<T>() as isize;
+        let mut i = 0;
+        while i + 16 <= len {
+            // Perform the swap 16 bytes at a time, `&mut` pointers never alias
+            ptr::copy_nonoverlapping(x.offset(i), t, 16);
+            ptr::copy_nonoverlapping(y.offset(i), x.offset(i), 16);
+            ptr::copy_nonoverlapping(t, y.offset(i), 16);
+            i += 16;
+        }
+        if i < len {
+            // Swap any remaining bytes
+            let rem = (len - i) as usize;
```

Comment:
I don't know if it ever matters, but it occurs to me that calculating `rem` as `len % 16` instead of `len - i` might be more obvious to an optimizer, as knowing what `len - i` is requires reasoning about a loop. Similarly, it would be better to test whether `rem == 0` than to test `i < len`, as it is more trivially determined at compile time.

I doubt this matters in most cases, but I could see the optimizer failing when the size is large, since the loop might not be fully unrolled.